### PR TITLE
MetaCPAN : Make better use of MetaCPAN API

### LINKED ISSUE #898: MetaCPAN - Switch to new API, show more data
Currently we are using an older API which only surfaces information for a given module. The API we're using is also case sensitive,  so "app::duckpan" will not show any results for "App::DuckPAN": https://duckduckgo.com/?q=cpanm+App%3A%3ADuckPAN, https://duckduckgo.com/?q=cpanm+app%3A%3Aduckpan.

It seems that the API has since been updated allowing for a general search endpoint:
http://api.metacpan.org/v0/release/_search?q=cpanminus

I suggest we switch to the new API to allow for true "search" of MetaCPAN Packages. In the event that the search does not exactly match a module name we could return a list of results and in the event that a module name is correctly indicated we should only show that result.

The API currently provides tons of info, most of which we probably don't need. You can choose which elements to return by indicating, for example `&fields=name,id` in the URL. `&fields=version,name,author,date,distribution` seems like a good start :)

E.g. http://api.metacpan.org/v0/release/_search?q=string%20trim&fields=name,abstract,version,metadata,author,download_url,date,distribution

This doc provides some insight into all the available fields:https://github.com/CPAN-API/cpan-api/blob/master/lib/MetaCPAN/Document/File.pm

https://duck.co/ia/view/meta_cpan

### PR DESCRIPTION
**Pull request by jbarrett**
###### What does your Pull Request do (check all that apply)?

Choose the most relevant items and use the following title template to name
your Pull Request.

- [ ] Improvement
    - [x] Enhancement: **`{MetaCPAN}: {Make better use of MetaCPAN API}`**

###### Description of changes

This is a Proof-of-Concept to evaluate POST Support for Spice:

https://github.com/duckduckgo/duckduckgo/pull/204
https://github.com/duckduckgo/p5-app-duckpan/pull/354

This is in no way ready for go-live, I was just checking everything was in place. This is here to help others do that too. Assuming all your project directories are in the same place you can checkout the branches above and...

```
$ perl -I../duckduckgo/lib/ -I../p5-app-duckpan/lib/ `which duckpan` server
```

###### Which issues (if any) does this fix?

fixes #898 - Switch to new API, show more data

This issue provided the impetus for all of these changes. The POST JSON blob used is basically [metacpan-web's module search](https://github.com/CPAN-API/metacpan-web/blob/848eec0/lib/MetaCPAN/Web/Model/API/Module.pm#L271-L411) - further sorting and filtering is probably required.

###### People to notify

@russellholt @MariagraziaAlastra @moollaza @jdorweiler

---

Instant Answer Page: https://duck.co/ia/view/meta_cpan

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by jbarrett** [comment]
Thanks @jdorweiler for deploying the supporting infrastructure for this on beta!

You should be able to [follow MetaCPAN's progress here](https://beta.duckduckgo.com/?q=cpan+midi)

File: share/spice/meta_cpan/meta_cpan.js [hunk 1]
```diff
@@ -2,40 +2,62 @@
...
+                sourceUrl: 'https://metacpan.org/search?size=50&search_type=modules&q=' + encodeURIComponent(query)
+            },
+            normalize: function(item) {
+                if ( !item.fields.module || !item.fields.module[0] ||
+                     ( !item.fields.description && !item.fields.abstract )
+                ) {
+                    return null;
+                }
+
+                var description = ( item.fields.description )
```
Comment (moollaza) [inline share/spice/meta_cpan/meta_cpan.js]: This should also work `var description = item.fields.description || item.fields.abstract`

File: share/spice/meta_cpan/meta_cpan.js [hunk 2]
```diff
@@ -2,40 +2,62 @@
...
+                    ? item.fields.description
+                    : item.fields.abstract;
+                var version = ( item.fields.version )
+                    ? item.fields.version
+                    : item.fields.module[0].version
+                var date = ( item.fields.date )
+                    ? new Date(item.fields.date).toLocaleDateString('en-US', {
```
Comment (moollaza) [inline share/spice/meta_cpan/meta_cpan.js]: We usually do this with moment.js:

```js
DDG.require('moment.js', function({
    Spice.add({
        ...
        var date = moment(item.fields.date).format("dd, MMM YYYY");
    });
});
```

**Comment by moollaza** [comment]
If possible, I'd like to merge + release the DDG changes before merging this so tests will continue to pass in the repo. Otherwise all PR's will start failing

**Comment by jbarrett** [comment]
Thanks for review @moollaza! Changes made.

Agreed, we absolutely have to have https://github.com/duckduckgo/duckduckgo/pull/204 in place before we consider merging this.

I think https://github.com/duckduckgo/p5-app-duckpan/pull/354 is less urgent as general breakage does not occur with the changes here - just the MetaCPAN spice fails to fetch the correct result.

**Comment by jbarrett** [comment]
OK, supporting infrastructure is in place, Travis looking good, this is basic but it's an improvement...

Wanna go ahead with this @jdorweiler @moollaza?

Thanks!

**Comment by jdorweiler** [comment]
:cake:

**Pull request merged**